cmd/cpe2cve/config.go

```diff
@@ -0,0 +1,300 @@
+}
+
+func (c *config) addFeedsFromArgs(args ...string) error {
+	if c.feeds == nil {
+		c.feeds = make(map[string][]string)
+	}
+	if len(args) > 0 && c.providerAt != 0 && c.provider == "" {
+		return fmt.Errorf("in order to output a provider, you need to specify a valid provider name using the -provider flag")
+	}
+	for _, file := range args {
+		c.feeds[c.provider] = append(c.feeds[c.provider], file)
+	}
+	return nil
+}
+
+func (c *config) validate() error {
+	if len(c.feeds) == 0 {
+		return fmt.Errorf("feed files weren't provided")
+	}
+	if c.cpesAt <= 0 {
+		return fmt.Errorf("-cpe flag wasn't provided")
+	}
+	if c.cvesAt <= 0 {
+		return fmt.Errorf("-cve flag wasn't provided")
+	}
+	if c.matchesAt < 0 {
+		return fmt.Errorf("-matches value is invalid %d", c.matchesAt)
+	}
+	if c.cwesAt < 0 {
+		return fmt.Errorf("-cwe value is invalid %d", c.cwesAt)
+	}
+	if c.cvss2at < 0 {
+		return fmt.Errorf("-cvss2 value is invalid %d", c.cvss2at)
+	}
+	if c.cvss3at < 0 {
+		return fmt.Errorf("-cvss2 value is invalid %d", c.cvss3at)
+	}
+	if c.cvssAt < 0 {
+		return fmt.Errorf("-cvss value is invalid %d", c.cvssAt)
+	}
+	return nil
+}
+
+type configFile struct {
+	CPEsAt             int
+	CVEsAt             int
+	MatchesAt          int
+	CWEsAt             int
+	CVSS2At            int
+	CVSS3At            int
+	CVSSAt             int
+	EraseFields        []int
+	ProviderAt         int
+	InFieldSeparator   string
+	InRecordSeparator  string
+	OutFieldSeparator  string
+	OutRecordSeparator string
+	NumProcessors      int
+	IndexDict          bool
+	CacheSize          int64
+	RequireVersion     bool
+	CPUProfile         string
+	MemoryProfile      string
+	FeedOverrides      []string
+	Feeds              map[string][]string
+}
+
+func readConfigFile(file string) (*configFile, error) {
+	f, err := os.Open(file)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	switch ext := path.Ext(file); ext {
+	case ".json":
+		return readJSONConfig(f)
+	case ".toml":
+		return readTOMLConfig(f)
+	default:
+		return nil, fmt.Errorf("unsupported file extension: %q", ext)
+	}
+}
+
+func readJSONConfig(r io.Reader) (*configFile, error) {
+	// Example:
+	// {
+	// 	"CvesAt": 3,
+	// 	"ProviderAt": 2,
+	// 	"Feeds": {
+	// 	  "foo": ["foo.json"],
+	// 	  "bar": ["bar.json", "bar2.json.gz"]
+	// 	 }
+	// }
+	var cf configFile
+	if err := json.NewDecoder(r).Decode(&cf); err != nil {
+		return nil, err
+	}
+	return &cf, nil
+}
+
+func readTOMLConfig(r io.Reader) (*configFile, error) {
+	// Example:
+	// CvesAt = 3
+	// ProviderAt = 2
+	// [Feeds]
+	// foo = ["foo.json"]
+	// bar = ["bar.json", "bar2.json.gz"]
+	var cf configFile
+	if _, err := toml.DecodeReader(r, &cf); err != nil {
+		return nil, err
+	}
+	return &cf, nil
+}
+
+// modify will override all values in config which are set to their respective defaults
+// in that way, command line flags have more priority over json config file
+func (cf *configFile) modify(c *config) {
```

Comment:
Do you have any specific use case for that in mind? To me making config file a priority is as reasonable as the opposite, but the former would allow to drop this function whatsoever.

Reply:
Sounds good, removed the use case